### ISSUE TITLE
Exhaustiveness checking?

### ISSUE THREAD

**Issue by AndrewMorsillo** [gvergnaud/ts-pattern#4]
Very cool library, thanks for creating it! Would it be possible to force exhaustiveness checking for union types? For example:

```
function getSomeString(unionType:  "foo" | "bar"): string{
     return match(unionType).with("foo", () => `A string result, but we won't check for "bar"!`).run()
}
```

This will not be a compile time error. If you write it as 

```
function getSomeString(unionType:  "foo" | "bar"): string{
    if(unionType === "foo"){
        return `A string result, but we won't check for "bar"!`
    } 
}
```

You get `TS2366: Function lacks ending return statement and return type does not include 'undefined'` because TS can verify that you haven't exhausted all of the cases.

It would be really nice if this library could take advantage of that. Failure to check all cases is a really easy bug to introduce!

**Comment by gvergnaud** [comment]
👋 Hey, Thanks for your feedback!

I had a go at implementing exhaustive pattern matching when I first wrote this library ([Here is the branch if you are curious](https://github.com/gvergnaud/ts-pattern/tree/exhaustive-patterns)). I agree having exhaustive patterns would be immensely useful, but it's actually pretty hard to get right.

Support `when` predicate doesn't make our lives easier to implement it, because if the function given isn't a [Type Guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types) there is no way to narrow the return type to something more precise. I think that's why exhaustive pattern matching isn't turned on by default on languages like Haskell.

The second issue we face is with literal patterns. Consider this case:

```ts
const input: number
match(input)
    .with(2, () => 'two')
    .with(3, () => 'three')
    .run()
```

This block will not compile, because `2` 's inferred type turns out to be `number`, so the second `.with` will have `Exclude<number, number>` which evaluates to `never` as narrowed input type and which isn't compatible with a `3` pattern. Of course you could do `.with(2 as const, ...)` to mitigate this issue, but the fact that the default behavior is incorrect and confusing discouraged me to actually push this experiment to the main branch. I looked for ways to force Typescript to consider the first argument of `.with` as `const` or `readonly` by default, but I didn't find a way make it work. 

Another approach we could have would be to make exhaustive patterns an opt-in behavior by extending the API surface of the library, as @oguimbal suggested in this [PR suggested](https://github.com/gvergnaud/ts-pattern/pull/6) but I'd like to make sure the proposed API makes sense before merging it.

**Comment by oguimbal** [comment]
That's why case guards have been removed from Elm pattern matching, if I remember right :)
And I guess that thats why they're not enabled by default in Haskell ... it can be quite confusing:

![image](https://user-images.githubusercontent.com/8973947/100478984-36d14980-30ed-11eb-83b3-ee49e098cc90.png)

**Comment by gvergnaud** [comment]
I opened a draft PR implementing exhaustive matching as an opt-in feature. Some info about the API and the tradeoffs in the description/comments. Feel free to give me a review! https://github.com/gvergnaud/ts-pattern/pull/7

**Comment by gvergnaud** [comment]
#7 has been merged adding opt-in exhaustive patterns to ts-pattern!! 🎉 

It's published on npm under the version `2.0.1-next.1` and it should work well for most usecases, feel free to try it and to tell me what you think!. The API looks like this:

```ts
type Option<a> = { kind: 'none' } | { kind: 'some'; value: a };

// This compiles:
match<Option<number>>({ kind: 'some', value: 3 })
        .exhaustive()
        .with({ kind: 'some' }, ({ value }) => value)
        .with({ kind: 'none' }, () => 0)
        .run();

// This doesn't
match<Option<number>>({ kind: 'some', value: 3 })
        .exhaustive()
        .with({ kind: 'some' }, ({ value }) => value)
        .run();
```

**Comment by gvergnaud** [comment]
Closing this issue since ts-pattern@2.1.0 has been released :)
